compiler/rustc_mir/src/const_eval/mod.rs

```diff
@@ -38,97 +40,159 @@ pub(crate) fn const_caller_location(
-        let fields = (0..n).map(|i| {
-            let field = ecx.mplace_field(&place, i).unwrap();
-            const_to_valtree_inner(ecx, &field)
-        });
-        // For enums, we preped their variant index before the variant's fields so we can figure out
-        // the variant again when just seeing a valtree.
-        let branches = variant.into_iter().chain(fields);
-        Some(ty::ValTree::Branch(
-            ecx.tcx.arena.alloc_from_iter(branches.collect::<Option<Vec<_>>>()?),
-        ))
+) -> EvalToValTreeResult<'tcx> {
+    let place = match variant {
+        Some(variant) => ecx.mplace_downcast(place, variant).unwrap(),
+        None => *place,
     };
+    let variant =
+        variant.map(|variant| Ok(Some(ty::ValTree::Leaf(ScalarInt::from(variant.as_u32())))));
+    let fields = (0..n).map(|i| {
+        let field = ecx.mplace_field(&place, i).unwrap();
+        const_to_valtree(ecx, &field)
+    });
+    // For enums, we preped their variant index before the variant's fields so we can figure out
+    // the variant again when just seeing a valtree.
+    let branches = variant.into_iter().chain(fields);
+    let branches = branches.collect::<Result<Option<Vec<_>>, _>>()?;
+    Ok(branches.map(|val| ty::ValTree::Branch(ecx.tcx.arena.alloc_from_iter(val))))
+}
+
+/// Convert an evaluated constant to a type level constant
+fn const_to_valtree<'tcx>(
+    ecx: &CompileTimeEvalContext<'tcx, 'tcx>,
+    place: &MPlaceTy<'tcx>,
+) -> EvalToValTreeResult<'tcx> {
     match place.layout.ty.kind() {
-        ty::FnDef(..) => Some(ty::ValTree::zst()),
+        ty::FnDef(..) => Ok(Some(ty::ValTree::zst())),
         ty::Bool | ty::Int(_) | ty::Uint(_) | ty::Float(_) | ty::Char => {
             let val = ecx.read_immediate(&place.into()).unwrap();
             let val = val.to_scalar().unwrap();
-            Some(ty::ValTree::Leaf(val.assert_int()))
+            Ok(Some(ty::ValTree::Leaf(val.assert_int())))
         }
 
         // Raw pointers are not allowed in type level constants, as we cannot properly test them for
         // equality at compile-time (see `ptr_guaranteed_eq`/`_ne`).
         // Technically we could allow function pointers (represented as `ty::Instance`), but this is not guaranteed to
         // agree with runtime equality tests.
-        ty::FnPtr(_) | ty::RawPtr(_) => None,
-        ty::Ref(..) => unimplemented!("need to use deref_const"),
+        ty::FnPtr(_) | ty::RawPtr(_) => Ok(None),
+        ty::Ref(..) => {
+            let mplace = ecx.deref_operand(&place.into()).unwrap();
+            if let Scalar::Ptr(ptr) = mplace.ptr {
+                assert_eq!(
+                    ecx.memory.get_raw(ptr.alloc_id).unwrap().mutability,
+                    Mutability::Not,
+                    "const_to_valtree cannot be used with mutable allocations as \
+                    that could allow pattern matching to observe mutable statics",
+                );
+            }
+
+            match mplace.meta {
```

Comment:
This is suspicious, usually it is a bug to `match` on `meta` like that -- instead one should match on the type (e.g. its unsized tail) and then assert that `meta` has the right shape.

Reply:
I did this transformation, much better